--- a/command/agent/job_endpoint.go
+++ a/command/agent/job_endpoint.go
@@ -1871,10 +1871,11 @@ func ApiAffinitiesToStructs(in []*api.Affinity) []*structs.Affinity {
 
 func ApiAffinityToStructs(a1 *api.Affinity) *structs.Affinity {
 	return &structs.Affinity{
-		LTarget: a1.LTarget,
-		Operand: a1.Operand,
-		RTarget: a1.RTarget,
-		Weight:  *a1.Weight,
+		LTarget:               a1.LTarget,
+		Operand:               a1.Operand,
+		RTarget:               a1.RTarget,
+		Weight:                *a1.Weight,
+		NormalizeNodeAffinity: *a1.NormalizeNodeAffinity,
 	}
 }
 
