--- a/nomad/structs/structs_test.go
+++ a/nomad/structs/structs_test.go
@@ -604,8 +604,8 @@ func TestJob_SpecChanged(t *testing.T) {
 		{
 			Name:     "With Affinities",
 			Changed:  false,
-			Original: &Job{Affinities: []*Affinity{{"A", "B", "=", 1}}},
-			New:      &Job{Affinities: []*Affinity{{"A", "B", "=", 1}}},
+			Original: &Job{Affinities: []*Affinity{{"A", "B", "=", 1, true}}},
+			New:      &Job{Affinities: []*Affinity{{"A", "B", "=", 1, true}}},
 		},
 	}
 
@@ -1155,14 +1155,14 @@ func TestJob_PartEqual(t *testing.T) {
 	}))
 
 	as := &Affinities{
-		&Affinity{"left0", "right0", "=", 0},
-		&Affinity{"left1", "right1", "=", 0},
-		&Affinity{"left2", "right2", "=", 0},
+		&Affinity{"left0", "right0", "=", 0, true},
+		&Affinity{"left1", "right1", "=", 0, true},
+		&Affinity{"left2", "right2", "=", 0, true},
 	}
 	require.True(t, as.Equal(&Affinities{
-		&Affinity{"left0", "right0", "=", 0},
-		&Affinity{"left2", "right2", "=", 0},
-		&Affinity{"left1", "right1", "=", 0},
+		&Affinity{"left0", "right0", "=", 0, true},
+		&Affinity{"left2", "right2", "=", 0, true},
+		&Affinity{"left1", "right1", "=", 0, true},
 	}))
 }
 
