diff --git a/api/tasks.go b/api/tasks.go
index 21d99bf4c2..236c4546e9 100644
--- a/api/tasks.go
+++ b/api/tasks.go
@@ -229,14 +229,16 @@ type Affinity struct {
 	RTarget string `hcl:"value,optional"`     // Right-hand target
 	Operand string `hcl:"operator,optional"`  // Constraint operand (<=, <, =, !=, >, >=), set_contains_all, set_contains_any
 	Weight  *int8  `hcl:"weight,optional"`    // Weight applied to nodes that match the affinity. Can be negative
+	NormalizeNodeAffinity *bool  `hcl:"normalize_node_affinity,optional"`
 }
 
-func NewAffinity(lTarget string, operand string, rTarget string, weight int8) *Affinity {
+func NewAffinity(lTarget string, operand string, rTarget string, weight int8, normalizeNodeAffinity bool) *Affinity {
 	return &Affinity{
 		LTarget: lTarget,
 		RTarget: rTarget,
 		Operand: operand,
 		Weight:  pointerOf(weight),
+		NormalizeNodeAffinity: pointerOf(bool(normalizeNodeAffinity)),
 	}
 }
 
@@ -244,6 +246,9 @@ func (a *Affinity) Canonicalize() {
 	if a.Weight == nil {
 		a.Weight = pointerOf(int8(50))
 	}
+	if a.NormalizeNodeAffinity == nil {
+		a.NormalizeNodeAffinity = pointerOf(bool(true))
+	}
 }
 
 func NewDefaultDisconnectStrategy() *DisconnectStrategy {
