--- a/version/version.go
+++ b/version/version.go
@@ -24,7 +24,7 @@ var (
 	// A pre-release marker for the version. If this is "" (empty string)
 	// then it means that it is a final release. Otherwise, this is a pre-release
 	// such as "dev" (in development), "beta", "rc1", etc.
-	VersionPrerelease = "dev"
+	VersionPrerelease = "rs-1"

 	// VersionMetadata is metadata further describing the build type.
 	VersionMetadata = ""
